addons/html_builder/static/src/builder/snippets_menu_tabs/invisible_elements_tab.js

```diff
@@ -0,0 +1,104 @@
+        // to an invisible snippet element is a list filled with all its
+        // descendant invisible snippets except those that have a closer
+        // invisible snippet ancestor.
+        const descendantPerSnippet = new Map();
+        // Filter the invisibleEls to only keep the root snippets
+        // and create the map ("descendantPerSnippet") of the snippets and
+        // their descendant snippets.
+        const rootInvisibleSnippetEls = invisibleEls.filter((invisibleSnippetEl) => {
+            const ancestorInvisibleEl = invisibleSnippetEl.parentElement.closest(
+                this.props.invisibleSelector
+            );
+            if (!ancestorInvisibleEl) {
+                return true;
+            }
+            const descendantSnippets = descendantPerSnippet.get(ancestorInvisibleEl) || [];
+            descendantPerSnippet.set(ancestorInvisibleEl, [
+                ...descendantSnippets,
+                invisibleSnippetEl,
+            ]);
+            return false;
+        });
+        // Insert an invisible snippet in its "parentEl" element.
+        const createInvisibleElement = (invisibleSnippetEl, isRootParent, isDescendant) => {
+            return {
+                snippetEl: invisibleSnippetEl,
+                name: getSnippetName(invisibleSnippetEl),
+                isRootParent,
+                isDescendant,
+                isVisible: invisibleSnippetEl.dataset.invisible !== "1",
+                children: [],
+            };
+        };
+        // Insert all the invisible snippets contained in "snippetEls" as
+        // well as their descendants in the "parentEl" element. If
+        // "snippetEls" is set to "rootInvisibleSnippetEls" and "parentEl"
+        // is set to "$invisibleDOMPanelEl[0]", then fills the right
+        // invisible panel like this:
+        // rootInvisibleSnippet
+        //     └ descendantInvisibleSnippet
+        //          └ descendantOfDescendantInvisibleSnippet
+        //               └ etc...
+        const createInvisibleElements = (snippetEls, isDescendant) => {
+            return snippetEls.map((snippetEl) => {
+                const descendantSnippetEls = descendantPerSnippet.get(snippetEl);
+                // An element is considered as "RootParent" if it has one or
+                // more invisible descendants but is not a descendant.
+                const invisibleElement = createInvisibleElement(
+                    snippetEl,
+                    !isDescendant && !!descendantSnippetEls,
+                    isDescendant
+                );
+                if (descendantSnippetEls) {
+                    invisibleElement.children = createInvisibleElements(descendantSnippetEls, true);
+                }
+                return invisibleElement;
+            });
+        };
+        this.state.invisibleEls = createInvisibleElements(rootInvisibleSnippetEls, false);
+    }
+
+    onInvisibleEntryClick(invisibleEntry) {
```

Comment:
`toggleElementVisibility(element)`?

Reply:
I agree with you for the method name but maybe `toggleElementVisibility(invisibleEntry)` instead of `toggleElementVisibility(element)` ?